src/plugins/data_source/server/data_source_service.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Client } from '@opensearch-project/opensearch';
+import LRUCache from 'lru-cache';
+import { Logger, SavedObjectsClientContract } from 'src/core/server';
+import { DataSourcePluginConfigType } from '../config';
+import { DataSourceClient } from './client';
+import { IDataSourceService } from './types';
+
+export class DataSourceService implements IDataSourceService {
+  private openSearchClientPool?: LRUCache<string, Client>;
+  private isClosed = false;
+
+  constructor(private logger: Logger, private config: DataSourcePluginConfigType) {}
+
+  public setup() {
+    const logger = this.logger;
+    const { size } = this.config.clientPool;
+
+    this.openSearchClientPool = new LRUCache({
+      max: size,
+      maxAge: 15 * 60 * 1000, // by default, TCP connection times out in 15 minutes
+
+      async dispose(endpoint, client) {
+        try {
+          await client.close();
+        } catch (error: any) {
+          // log and do nothing since we are anyways evicting the client object from cache
+          logger.warn(
+            `Error closing OpenSearch client when removing from client pool: ${error.message}`
+          );
+        }
+      },
+    });
+    this.logger.info(`Created data source client pool of size ${size}`);
+  }
+
+  public getCachedClient(endpoint: string) {
+    return this.openSearchClientPool!.get(endpoint);
+  }
+
+  public addClientToPool(endpoint: string, client: Client) {
+    this.openSearchClientPool!.set(endpoint, client);
+  }
+
+  getDataSourceClient(logger: Logger, savedObjectClient: SavedObjectsClientContract) {
+    return new DataSourceClient({
+      logger,
+      dataSourceService: this,
+      scopedSavedObjectsClient: savedObjectClient,
+      config: this.config,
+    });
+  }
```

Comment:
data source client and data source service relationship not clear to me. it seems client relies on the service, however, exposing this interface from the low level service class and letting service to pass itself (`this`) to client is kind of weird

Reply:
Move logic from datasource client to datasource service, and expose functional interfaces only